pkg/apis/core/v1beta1/helper/helper.go

```diff
@@ -541,14 +541,40 @@ func SeedSettingOwnerChecksEnabled(settings *gardencorev1beta1.SeedSettings) boo
 	return settings == nil || settings.OwnerChecks == nil || settings.OwnerChecks.Enabled
 }
 
-// SeedSettingDependencyWatchdogEndpointEnabled returns true if the depedency-watchdog-endpoint is enabled.
-func SeedSettingDependencyWatchdogEndpointEnabled(settings *gardencorev1beta1.SeedSettings) bool {
-	return settings == nil || settings.DependencyWatchdog == nil || settings.DependencyWatchdog.Endpoint == nil || settings.DependencyWatchdog.Endpoint.Enabled
+// SeedSettingDependencyWatchdogWeederEnabled returns true if the dependency-watchdog-weeder is enabled. By default, this will be enabled unless explicitly set to false in `settings.DependencyWatchdog.Weeder.Enabled` or `settings.DependencyWatchdog.Endpoint.Enabled`(deprecated).
+// TODO:(himanshu-kun) Rely just on `settings.DependencyWatchdog.Weeder` field after few releases. For now `Weeder` field is preferred over `Endpoint` field, if both are set explicitly.
+func SeedSettingDependencyWatchdogWeederEnabled(settings *gardencorev1beta1.SeedSettings) bool {
+	if settings != nil && settings.DependencyWatchdog != nil {
```

Comment:
you could have implemented it as follows: 
```golang
func SeedSettingDependencyWatchdogWeederEnabled(settings *gardencorev1beta1.SeedSettings) bool {
         weederEnabled := settings.DependencyWatchdog.Weeder == nil || settings.DependencyWatchdog.Weeder.Enabled
         // settings.DependencyWatchdog.Endpoint is deprecated and will be removed in the future release.
         endpointEnabled := settings.DependencyWatchdog.Endpoint == nil || settings.DependencyWatchdog.Endpoint.Enabled
    	return settings == nil ||
		settings.DependencyWatchdog == nil ||
                 weederEnabled || endpointEnabled
}
```

Reply:
This has 2 problems:
* `settings.DependencyWatchdog.Weeder == nil` would return nil pointer error if `settings.DependencyWatchdog=nil` , but that could be dealt with.
* if we have set `settings.DependencyWatchdog.Weeder.Enabled` to `false` then we don't want weeder to be deployed,and because the other remains unset (lets say `endpoint` is unset .i.e. nil) as we don't allow to set both, then `endpointEnabled` would become `true` , now the function would return `true`.

So question , should I change the return to 
```code
return settings == nil ||
		settings.DependencyWatchdog == nil ||
                 (weederEnabled && endpointEnabled)
```
because we are sure that both are not setting together.

Reply:
The original method has a reverse logic because it assumes that if nothing is set then default value will be enabled. That is why if DependencyWatchDog is nil or settings is nil or Endpoint | Weeder is nil then it is assumed that it is enabled. This assumption of default = true makes consumption easy but complicates the check.

>because we are sure that both are not setting together.

We can only be sure if we validate this and reject if both are set and ensuring that for new clusters only weeder should be set. So is this is true then the above check should be fine.

Reply:
Yes I have introduced changes regarding that validation in the latest commit
In that case I am making my proposed changes.

Reply:
made changes 3fc186b78572f106350c6aaf90932b4d83afa801